pkg/globalnet/controllers/uninstall.go

```diff
@@ -0,0 +1,178 @@
+	routeAgent "github.com/submariner-io/submariner/pkg/routeagent_driver/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes/scheme"
+	rest "k8s.io/client-go/rest"
+	"k8s.io/klog"
+	utilexec "k8s.io/utils/exec"
+)
+
+func UninstallDataPath() {
+	ipt, err := iptables.New()
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	natTableChains := []string{
+		// The chains have to be deleted in a specific order.
+		constants.SmGlobalnetEgressChainForCluster,
+		constants.SmGlobalnetEgressChainForHeadlessSvcPods,
+		constants.SmGlobalnetEgressChainForNamespace,
+		constants.SmGlobalnetEgressChainForPods,
+		constants.SmGlobalnetIngressChain,
+		constants.SmGlobalnetMarkChain,
+		constants.SmGlobalnetEgressChain,
+	}
+
+	for _, chain := range natTableChains {
+		err = ipt.FlushIPTableChain(constants.NATTable, chain)
+		if err != nil {
+			// Just log an error as this is part of uninstallation.
+			klog.Error(err)
+		}
+	}
+
+	err = ipt.FlushIPTableChain(constants.NATTable, routeAgent.SmPostRoutingChain)
+	if err != nil {
+		klog.Error(err)
+	}
+
+	if err := ipt.DeleteIPTableRule(constants.NATTable, "PREROUTING", constants.SmGlobalnetIngressChain); err != nil {
+		klog.Errorf("error deleting iptables rule for %q in PREROUTING chain: %v\n", constants.SmGlobalnetIngressChain, err)
+	}
+
+	for _, chain := range natTableChains {
+		err = ipt.DeleteIPTableChain(constants.NATTable, chain)
+		if err != nil {
+			klog.Error(err)
+		}
+	}
+
+	ipsetIface := ipset.New(utilexec.New())
+
+	ipSetList, err := ipsetIface.ListSets()
+	if err != nil {
+		klog.Errorf("Error listing ipsets: %v", err)
+	}
+
+	for _, set := range ipSetList {
+		if strings.HasPrefix(set, IPSetPrefix) {
+			err = ipsetIface.DestroySet(set)
+			if err != nil {
+				klog.Errorf("Error destroying the ipset %q: %v", set, err)
+			}
+		}
+	}
+}
+
+func DeleteGlobalnetObjects(smClientSet *versioned.Clientset, cfg *rest.Config) {
+	err := smClientSet.SubmarinerV1().ClusterGlobalEgressIPs("").DeleteCollection(context.TODO(), metav1.DeleteOptions{},
+		metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Error deleting the clusterGlobalEgressIPs: %v", err)
+	}
+
+	err = smClientSet.SubmarinerV1().GlobalEgressIPs("").DeleteCollection(context.TODO(), metav1.DeleteOptions{},
+		metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Error deleting the globalEgressIPs: %v", err)
+	}
+
+	deleteGlobalIngressIPs(smClientSet, cfg)
+}
+
+func deleteGlobalIngressIPs(smClientSet *versioned.Clientset, cfg *rest.Config) {
+	defer deleteAllGlobalIngressIPObjs(smClientSet)
+
+	restMapper, err := util.BuildRestMapper(cfg)
+	if err != nil {
+		klog.Errorf("Error creating the RestMapper: %v", err)
+		return
+	}
+
+	_, gvr, err := util.ToUnstructuredResource(&corev1.Service{}, restMapper)
+	if err != nil {
+		klog.Errorf("Error converting resource: %v", err)
+		return
+	}
+
+	dynClient, err := dynamic.NewForConfig(cfg)
```

Comment:
I think we should pass in the dynamic client for consistency (we pass in the submariner client) and also to facilitate unit tests. Also, since we're using it to delete `Services`, we could just use a kube client.

Reply:
I want to re-use the existing methods like getService, finalizer.Remove and deleteService which work on dynamic resources. Hence passing dynamic client instead of kubeclient.

Reply:
We should still pass in `dynamic.Interface` to facilitate unit tests but we can change that later (although it would be nice to add unit tests with this PR 😄 ).